mantidproject/mantid
# added abillity to set a scale for stl files to LoadSampleShape and Environment

**Description of work.**
added a scale option to `LoadSampleShape` and `LoadSampleEnvironment` because the units in an stl file are arbitrary, the units were always assumed to be metres, which may not be the case, the scale currently accepts metres, centimetres and milimetres in the format "m", "cm", "mm"


**To test:**
run the following script, changing the input files to your own test data
```
Load(Filename='/home/sjenkins/Work/Build-1/ExternalData/Testing/Data/SystemTest/WISH/input/18_1/WISH00040503.raw', OutputWorkspace='ws',SpectrumList="100")

ConvertUnits(InputWorkspace="ws",Target="Wavelength", OutputWorkspace="ws1")
xml='''<cylinder id="A">
  <centre-of-bottom-base x="0.0" y="0.0" z="-0.15" />  
  <axis x="0.0" y="0.0" z="1" />
  <radius val="0.025" />
  <height val="0.3" />
</cylinder>'''
SetSample("ws1",Geometry={"Shape":"CSG","Value":xml}, Material={"ChemicalFormula":"V","SampleNumberDensity":"0.1" })
SetSampleMaterial(InputWorkspace="ws1",ChemicalFormula="V",SampleNumberDensity=0.1)
MonteCarloAbsorption(InputWorkspace="ws1",OutputWorkspace="ws1absorb",NumberOfWavelengthPoints=50)

ConvertUnits(InputWorkspace="ws",Target="Wavelength", OutputWorkspace="ws2")
LoadSampleShape(InputWorkspace="ws2",Filename="/home/sjenkins/Work/Build-1/ExternalData/Testing/Data/UnitTest/cylinder.stl",Scale="cm",OutputWorkspace="ws2")
SetSampleMaterial(InputWorkspace="ws2",ChemicalFormula="V",SampleNumberDensity=0.1)
MonteCarloAbsorption(InputWorkspace="ws2",OutputWorkspace="ws2absorb",NumberOfWavelengthPoints=50)
```
plot ws1absorb against ws2absorb, they should be fairly similar (the differences are due to stl not being as accurate a representation of a cylinder)

rerun the script but change `LoadSampleShape(InputWorkspace="ws2",Filename="/home/sjenkins/Work/Build-1/ExternalData/Testing/Data/UnitTest/cylinder.stl",Scale="cm",OutputWorkspace="ws2")` to have "m" or "mm" instead of "cm" ws2absorb should change dramatically getting much larger than ws1absorb for mm and much smaller for m.

Fixes #25500

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
